File: app/src/test/java/com/mohsenoid/rickandmorty/data/RepositoryQueryCharactersTest.java
package com.mohsenoid.rickandmorty.data;

import org.json.JSONException;
import org.junit.Test;

import java.io.IOException;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

public class RepositoryQueryCharactersTest extends RepositoryTest {

    @Test
    public void testIfQueryCharactersCallsApiClientMethodWhenIsOnline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(true);

        // WHEN
        repository.queryCharactersByIds(null, null);

        // THEN
        verify(db, times(1)).queryCharactersByIds(null);
        verify(networkClient, times(1)).getCharactersByIds(null);
    }

    @Test
    public void testIfQueryCharactersCallsApiClientMethodWhenIsOffline() throws IOException, JSONException {
        // GIVEN
        stubConfigProviderIsOnline(false);

        // WHEN
        repository.queryCharactersByIds(null, null);

        // THEN
        verify(db, times(1)).queryCharactersByIds(null);
        verify(networkClient, times(0)).getCharactersByIds(null);
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/data/RepositoryQueryCharactersTest.kt
package com.mohsenoid.rickandmorty.data

import com.mohsenoid.rickandmorty.test.DataFactory
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import kotlinx.coroutines.runBlocking
import org.junit.Test

class RepositoryQueryCharactersTest : RepositoryTest() {

    @Test
    fun `test if queryCharactersByIds calls networkClient when isOnline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(true)
            val characterIds = DataFactory.randomIntList(5)

            // WHEN
            repository.queryCharactersByIds(characterIds, null)

            // THEN
            verify(characterDao, times(1)).queryCharactersByIds(characterIds)
            verify(networkClient, times(1)).getCharactersByIds(characterIds)
        }
    }

    @Test
    fun `test if queryCharactersByIds calls db only when isOffline`() {
        runBlocking {
            // GIVEN
            stubConfigProviderIsOnline(false)
            val characterIds = DataFactory.randomIntList(5)

            // WHEN
            repository.queryCharactersByIds(characterIds, null)

            // THEN
            verify(characterDao, times(1)).queryCharactersByIds(characterIds)
            verify(networkClient, times(0)).getCharactersByIds(characterIds)
        }
    }
}
